barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -509,4 +514,113 @@ namespace avm_alu(256);
     // 128-bit multiplication and CAST need two rows in ALU trace. We need to ensure
     // that another ALU operation does not start in the second row.
     #[TWO_LINE_OP_NO_OVERLAP]
-    (op_mul * ff_tag + op_cast) * alu_sel' = 0;
\ No newline at end of file
+    (op_mul * ff_tag + op_cast) * alu_sel' = 0;
+
+    // ========= SHIFT LEFT/RIGHT OPERATIONS ===============================
+    // Given (1) an input b, within the range [0, 2**128-1], 
+    //       (2) a value s, the amount off bits to shift b by,
+    //       (3) and a memory tag, mem_tag that supports a maximum of t bits.
+    // Split input b into Big Endian hi and lo limbs, (we re-use the b_hi and b_lo columns we used for the comparison operators)
+    // b_hi and b_lo, and the number of bits represented by the memory tag, t.
+    // QUESTION: SHOULD S BE CONSTRAINED TO BE U8 -> i.e. when would we shift more than 255 bits when the max number of bits of a is 128bits?
+    // If we are shifting by more than the bit length represented by the memory tag, the result is trivially zero
+    //
+    // === Steps when performing SHR
+    // (1) Prove the correct decomposition: b_hi * 2**s + b_lo = b
+    // (2) Range check b_hi < 2**(t-s) && b_lo < 2**s, ensure we have not overflowed the limbs during decomp
+    // (3) Return b_hi
+    //
+    //  <--(t-s) bits --> |   <-- s bits -->
+    // -------------------|-------------------
+    // |      b_hi        |       b_lo       | --> b
+    // ---------------------------------------
+    //
+    // === Steps when performing SHL
+    // (1) Prove the correct decomposition: b_hi * 2**(t-s) + b_lo = b
+    // (2) Range check b_hi < 2**s && b_lo < 2**(t-s)
+    // (3) Return b_lo * 2**s
+    // 
+    //  <-- s bits -->   | <-- (t-s) bits -->
+    // ------------------|-------------------
+    // |      b_hi       |      b_lo        | --> b
+    // --------------------------------------
+
+    // TODO: Possibly optimised with variable length checks
+    // Check that b_lo and b_hi are range checked such that:
+    // SHR: b_hi < 2**(t - s) && b_lo < 2**s
+    // SHL: b_hi < 2**s && b_lo < 2**(t - s)
+
+    // In lieu of a variable length check, we can utilise 2 fixed range checks instead.
+    // Given the dynamic range check of 0 <= b_hi < 2**(t-s), where s < t
+    // (1) 0 <= b_hi < 2**t
+    // (2) 0 <= 2**(t - s) - a_hi < 2**t
+    // Note that (1) is guaranteed elsewhere through the tagged memory model, so we focus on (2) here. 
+
+    // === General Notes:
+    // There are probably ways to merge various relations for the SHL/SHR, but they are separate
+    // now while we are still figuring out.
+
+    
+    // We re-use the a_lo and a_hi cols from the comparison operators the range checks
+    // (1) a_lo = 2**b - b_lo, (2) a_hi = 2**(t-b) - b_hi
+    // === Range checks: (1) a_lo < 2**128, (2) a_hi < 2**128.
+    #[SHR_RANGE_0]
+    shift_lt_bit_len * op_shr * (a_lo - (two_pow_b - b_lo)) = 0;
```

Comment:
I think we have an off-by-one error. We want b_lo strictly smaller than two_pow_s. If they would be equal, then a_lo is zero and passes the range check which is not ok.
All the range checks need to be adjusted I think with like this:
a_lo = 2**b - b_lo - 1